pkg/api/graph/deployment.go

```diff
@@ -0,0 +1,415 @@
+		group := ServiceGroup{}
+
+		matches := NodesByKind(other, c, ServiceGraphKind, DeploymentConfigGraphKind)
+		svcs, dcs, _ := matches[0], matches[1], matches[2]
+
+		for _, n := range svcs {
+			covers := []*DeploymentConfigNode{}
+			for _, neighbor := range other.Neighbors(n) {
+				switch other.EdgeKind(g.EdgeBetween(neighbor, n)) {
+				case ExposedThroughServiceGraphEdgeKind:
+					covers = append(covers, neighbor.(*DeploymentConfigNode))
+				}
+			}
+			group.Services = append(group.Services, ServiceReference{
+				Service: n.(*ServiceNode),
+				Covers:  covers,
+			})
+		}
+		sort.Sort(SortedServiceReferences(group.Services))
+
+		for _, n := range dcs {
+			d := n.(*DeploymentConfigNode)
+			group.Deployments = append(group.Deployments, DeploymentFlow{
+				Deployment: d,
+				Images:     deploys[d],
+			})
+		}
+		sort.Sort(SortedDeploymentFlows(group.Deployments))
+
+		if len(dcs) == 0 || len(svcs) == 0 {
+			unknown := g.SubgraphWithNodes(c, ExistingDirectEdge)
+			for _, n := range unknown.NodeList() {
+				g.PredecessorEdges(n, AddGraphEdgesTo(unknown), BuildOutputGraphEdgeKind)
+			}
+			unknown = unknown.EdgeSubgraph(ReverseGraphEdge)
+			for _, n := range unknown.RootNodes() {
+				if flow, ok := ImageDeploymentFlowFromNode(unknown, n, make(NodeSet)); ok {
+					group.Builds = append(group.Builds, flow)
+				}
+			}
+		}
+		sort.Sort(SortedImageDeploymentFlows(group.Builds))
+
+		serviceGroups = append(serviceGroups, group)
+	}
+	sort.Sort(SortedServiceGroups(serviceGroups))
+	return serviceGroups
+}
+
+// UncoveredDeploymentFlowEdges preserves (and duplicates) edges that were not
+// covered by a deployment flow. As a special case, it preserves edges between
+// Services and DeploymentConfigs.
+func UncoveredDeploymentFlowEdges(covered NodeSet) EdgeFunc {
+	return func(g Interface, head, tail graph.Node, edgeKind int) bool {
+		if edgeKind == ExposedThroughServiceGraphEdgeKind {
+			return AddReversedEdge(g, head, tail, ReferencedByGraphEdgeKind)
+		}
+		if covered.Has(head.ID()) && covered.Has(tail.ID()) {
+			return false
+		}
+		return AddReversedEdge(g, head, tail, ReferencedByGraphEdgeKind)
+	}
+}
+
+// UncoveredDeploymentFlowNodes includes nodes that either services or deployment
+// configs, or which haven't previously been covered.
+func UncoveredDeploymentFlowNodes(covered NodeSet) NodeFunc {
+	return func(g Interface, node graph.Node) bool {
+		switch node.(type) {
+		case *DeploymentConfigNode, *ServiceNode:
+			return true
+		}
+		return !covered.Has(node.ID())
+	}
+}
+
+// ServiceReference is a service and the DeploymentConfigs it covers
+type ServiceReference struct {
+	Service *ServiceNode
+	Covers  []*DeploymentConfigNode
+}
+
+// ServiceGroup is a related set of resources that should be displayed together
+// logically. They are usually sorted internally.
+type ServiceGroup struct {
+	Services    []ServiceReference
+	Deployments []DeploymentFlow
+	Builds      []ImageDeploymentFlow
```

Comment:
so the same ImageDeploymentFlows referenced by Builds might also be referenced by ImageDeploymentFlows within the DeploymentFlows under Deployments?  Or are the ImageDeploymentFlows referenced by Builds only ones that are not already part of a DeploymentFlow?